8-channel USB audio output

h4yn0nnym0u5e opened the pull request:
Brief testing suggests this works, but it's not optimised and may have glitches and such.

Still giving me "credit" for others' work, and refusing to merge automatically. Grrr. Ahhh ... OK ... removed some 'scope trigger code and it's OK to merge now ... then put in the "improved sample count calculation", and a conflict has appeared!

h4yn0nnym0u5e committed:
Initial copy 2022-11-26

h4yn0nnym0u5e committed:
Make 8-channel audio output work

Not at all optimised at the moment...

h4yn0nnym0u5e committed:
Fixes conflicts? Let's see...

h4yn0nnym0u5e committed:
Improve sample count calculation

Will be needed for sample rates other than 44.1kHz

mcginty commented:
Ah yeah, it's because you're not rebased on top of the `master` for this repository, you're rebased on top of the parent repo's `master`. No worries, not that hard to fix after.

h4yn0nnym0u5e commented:
Yup - the ways of git are strange indeed ... would it make life easier for you if I spin up a new github account just for this project, so I can fork directly off your repos?

h4yn0nnym0u5e committed:
Implement higher channel count USB classes

These are only defined if the settings in usb_desc.h permit, in particular the value of AUDIO_CHANNELS. For example, if an attempt is made to use an 8-channel object when only 6 channels are configured, then a compile-time error will result.

mcginty commented:
@h4yn0nnym0u5e You can accomplish the same thing with just adding my repository as another remote (`git remote add mcginty https://github.com/mcginty/teensy-cores`), and then when developing make your branch by doing something like `git checkout -b feature/my-feature mcginty/master` so that it's based on this repository's master. Then if you had other changes in the works for the upstream master, there's no issues with your other branches :).

h4yn0nnym0u5e commented:
I'll give that a go ... at the moment I'm in a bit of a tangle because I tried to test 48kHz, had to make changes to get it working, and I'm now distinctly confused! It did eventually work, though I had to give Windows a good kicking to forget the old 44.1kHz capability and recognise the updated descriptor.

If you think this PR is vaguely OK, and pull it, then I can try to start again from your updated master with most of my changes already in place.

mcginty commented on teensy4/usb_audio.cpp:
```diff
@@ -94,7 +92,7 @@ void usb_audio_configure(void)
 	usb_audio_underrun_count = 0;
 	usb_audio_overrun_count = 0;
-	feedback_accumulator = 739875226; // 44.1 * 2^24
+	feedback_accumulator = (AUDIO_SAMPLE_RATE_EXACT / 1000.0f) * 0x1000000; // samples/millisecond * 2^24
```
To keep it consistent with the previous hardcoded number, should probably round up or at least add 1:

```suggestion
	feedback_accumulator = (AUDIO_SAMPLE_RATE_EXACT / 1000.0f) * 0x1000000 + 1; // samples/millisecond * 2^24
```

h4yn0nnym0u5e replied:
Hmmm ... good point, but as it's multiplied by 2^24 an "error" of one count isn't going to matter much, and presumably it'll promptly get corrected as it's trying to match the PC's demand rate to the Teensy clock. Or something. I'm out of my depth here!

I have checked, and the cast to integer happens late enough that the 0.1 doesn't get lost. There is (maybe) a bit of loss of precision, I get 739875200 printed out, but not 100% sure the adjustments to the value haven't started by then.

mcginty replied:
> Hmmm ... good point, but as it's multiplied by 2^24 an "error" of one count isn't going to matter much, and presumably it'll promptly get corrected as it's trying to match the PC's demand rate to the Teensy clock. Or something. I'm out of my depth here!

Agreed, just wanted to at least note the difference in case there was some math or concept that depended on that number :).

> I have checked, and the cast to integer happens late enough that the 0.1 doesn't get lost. There is (maybe) a bit of loss of precision, I get 739875200 printed out, but not 100% sure the adjustments to the value haven't started by then.

Sorry, I'm not sure I understand what the 0.1 you're referring to is. You get the 739875200 after the code has been running and the accumulator's initial value has been changed I take it?

<img width="923" alt="Screenshot 2022-11-28 at 16 02 33" src="https://user-images.githubusercontent.com/373823/204390323-b97ff6db-a23e-4884-acc5-b9a71a261926.png">

h4yn0nnym0u5e replied:
I'm assuming so ... 44100 / 1000 = 44.1 ... if the integer cast happen after this calculation, we'd get 44 * 2^24 and the fractional part, 0.1, would be "lost", as I put it. Actually, I think the calculation is done as float, not double, so it's actually probably a loss of precision. Still not really important!

mcginty commented on teensy4/usb_audio.cpp:
```diff
@@ -323,114 +323,168 @@ static void copy_from_buffers(uint32_t *dst, int16_t *left, int16_t *right, unsi
+			if (NULL == chans[i]) // sent NULL: make implied silence into real data
+			{
+				chans[i] = allocate();
+				if (NULL != chans[i])
+					memset(chans[i]->data, 0, sizeof(chans[i]->data));
+				else
+					break; // no block available, exit early
```
```suggestion
				if (NULL != chans[i]) {
					memset(chans[i]->data, 0, sizeof(chans[i]->data));
				} else {
					for (int j=0; i<AUDIO_CHANNELS; j++) {
						if (NULL != chans[j]) release(chans[j]);
					};
					return; // no block available, exit early
				}
```

Personally, something like this feels more obvious to me (along with getting rid of that `if (i >= AUDIO_CHANNELS) { ... } else { ... }` block), just as far as flow legibility is concerned.

h4yn0nnym0u5e replied:
I have a serious dislike of the "early return" style of error handling! I find it tends to make the code hard to maintain or modify, because you have to catch _all_ the instances and figure out if anything needs unwinding before the return, e.g. a disabled interrupt. It also make it hard to test: I put pin toggles at the start and end of usb_audio_receive_callback(), and they didn't work as expected because of all the early returns - in order to instrument it on my 'scope I ended up wrapping it in another function...

Using the final value of i is a bit obscure, I'll admit; I should have created an "all OK" flag and used that, really.

mcginty replied:
I would absolutely tend to agree with you, coming from mainly doing Rust myself. Unfortunately the rest of this codebase doesn't, and maybe the answer is to simply have that philosophy for this USB audio class. Be the change you want to see in the world ;). Please ignore this suggestion, and I can make the rest of the code consistent with this rule.

h4yn0nnym0u5e replied:
Too true ... and I don't feel the need to change it unless I'm doing heavy revisions, I just sigh and move on...

mcginty commented on teensy4/usb_audio.cpp:
```diff
@@ -323,114 +323,168 @@ static void copy_from_buffers(uint32_t *dst, int16_t *left, int16_t *right, unsi
 
+static void interleave_from_blocks(int16_t* transmit_buffer,	//!< next free sample in USB transmit buffer
+								audio_block_t** outgoing, //!< array of pointers to source audio blocks
+								int chans, //!< number of entries in the array
```
nitpick: any reason for this to not just use the AUDIO_CHANNELS constant since it's a privately used function for this class?

h4yn0nnym0u5e replied:
Only a vague thought that this interleaving code is fairly common in the library, and should at some point be pulled out into a super-optimised utility function, which would usually cater for several different "channel" counts.

mcginty replied:
Fair enough! I don't mind too much either way :).

mcginty reviewed (commented):
This looks excellent! I really appreciate your comments in the code, and the general simplicity you've managed in this diff.

mcginty commented:
@h4yn0nnym0u5e since this PR is most definitely "vaguely OK" :wink: I'll go ahead and merge it, leaving these comments as future notes for your later PR (or if you agree with them, I can just throw it in a separate commit as well if you want to take your time with the other PR).

mcginty commented:
Merged as https://github.com/mcginty/teensy-cores/commit/d2039c22ef44a056539b6fdb446c8babeec07dd2.
